### ISSUE TITLE
Browsing some users' files seems broken

### ISSUE THREAD

**Issue by steinuil** [jpdillingham/Soulseek.NET#503]
When I try to browse some users' files from slskd, it seems to download the whole response and then throws this exception:

```
Soulseek.SoulseekClientException: Failed to browse user red_book: The peer returned an invalid browse response
 ---> Soulseek.MessageReadException: The peer returned an invalid browse response
 ---> Soulseek.MessageCompressionException: Failed to decompress the message payload
 ---> Soulseek.Messaging.Compression.ZStreamException: inflating:
   at Soulseek.Messaging.Compression.ZOutputStream.Write(Byte[] b1, Int32 off, Int32 len)
   at Soulseek.Messaging.MessageReader`1.<Decompress>g__CopyStream|33_0(Stream input, Stream output)
   at Soulseek.Messaging.MessageReader`1.Decompress(Byte[] inData, Byte[]& outData)
   --- End of inner exception stack trace ---
   at Soulseek.Messaging.MessageReader`1.Decompress(Byte[] inData, Byte[]& outData)
   at Soulseek.Messaging.MessageReader`1.Decompress()
   at Soulseek.Messaging.Messages.BrowseResponseFactory.FromByteArray(Byte[] bytes)
   at Soulseek.Messaging.Handlers.PeerMessageHandler.HandleMessageRead(Object sender, Byte[] message)
   --- End of inner exception stack trace ---
   at Soulseek.SoulseekClient.BrowseInternalAsync(String username, BrowseOptions options, CancellationToken cancellationToken)
   --- End of inner exception stack trace ---
   at Soulseek.SoulseekClient.BrowseInternalAsync(String username, BrowseOptions options, CancellationToken cancellationToken)
   at slskd.API.Controllers.UserController.Browse(String username)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor.TaskOfIActionResultExecutor.Execute(IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object[] arguments)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeActionMethodAsync>g__Awaited|12_0(ControllerActionInvoker invoker, ValueTask`1 actionResultValueTask)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeNextActionFilterAsync>g__Awaited|10_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Rethrow(ActionExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeInnerFilterAsync>g__Awaited|13_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResourceFilter>g__Awaited|24_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResourceExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeFilterPipelineAsync>g__Awaited|19_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)
   at Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)
   at Microsoft.AspNetCore.Authorization.Policy.AuthorizationMiddlewareResultHandler.HandleAsync(RequestDelegate next, HttpContext context, AuthorizationPolicy policy, PolicyAuthorizationResult authorizeResult)
   at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at Serilog.AspNetCore.RequestLoggingMiddleware.Invoke(HttpContext httpContext)
```

Then again, I tried browsing this user from SoulseekQt and it also seems to choke on the response.

Judging from the time it takes to download the responses, maybe it has something to do with the size of the file list. I found [this post on reddit](https://old.reddit.com/r/Soulseek/comments/8lv8p8/not_being_able_to_browse_users_files/) about this problem, so maybe it's SoulseekQt that has trouble serializing very big responses.

**Comment by mathiascode** [comment]
Not entirely sure why SoulseekQt chokes on the response, but it is rather large (10TB of shared files). Nicotine+ displays it fine, although it takes around a minute to retrieve it.

**Comment by jpdillingham** [comment]
I can confirm this is a problem.  It had been in the back of my mind to find someone with a huge share and test this, so thanks for identifying one for me.

Nicotine+ does indeed handle this properly, so there's either something different in the way I'm handling the response message, or there's a bug in the zlib code I'm using for compression.

**Comment by jpdillingham** [comment]
Wireshark shows a _ton_ of retransmitted and out-of-order packets while sniffing the response data from this user.  I'm guessing that the integrity of the response is being compromised in flight.

It isn't immediately clear to me why Nicotine+ (and Python) is capable of putting the packets back together properly and .NET doesn't seem to be.  The problem is in the TCP code _somewhere_, it'll take some effort to figure it out.

**Comment by steinuil** [comment]
I'll make sure to post it here if I find another user that triggers this bug then.

**Comment by jpdillingham** [comment]
#504 fixes the issue for this user and I suspect most or all others, but it's kind of sketchy so please continue to keep an eye on it.

I don't know enough about the zlib algorithm to comment intelligently, but reading past a buffer seems like an error in logic so increasing the buffer size is probably kicking the can.

The network issues were a red herring (thankfully), I performed the browse again shortly after posting and it completed with no errors in Wireshark but still failed.

**Comment by jpdillingham** [comment]
I should add that the API struggles to serialize the response to json and return it in a timely manner, and the web app definitely won't be able to display it.  Those are shortcomings of the example app/slskd and will be addressed in the future.

**Comment by jpdillingham** [comment]
Fixed in 2.2.0; I'll get this over to slskd and released tonight or tomorrow.
